File: src/Services/Totem.Application/Services/ServiceTypeServices/ServiceTypeService.cs
using FluentValidation;
using Totem.Common.Domain.Notification;
using Totem.Common.Localization.Resources;
using Totem.Common.Services;
using Totem.Domain.Aggregates.ServiceTypeAggregate;
using Totem.Domain.Models.ServiceTypeModels;

namespace Totem.Application.Services.ServiceTypeServices
{
	public class ServiceTypeService : BaseService, IServiceTypeService
	{
		private readonly IServiceTypeQueires _queries;
		private readonly IServiceTypeRepository _repository;

		public ServiceTypeService(INotificator notificator, IServiceTypeQueires serviceTypeQueires, IServiceTypeRepository serviceTypeRepository) : base(notificator)
		{
			_queries = serviceTypeQueires;
			_repository = serviceTypeRepository;
		}

		public async Task<(Result result, Guid data)> CreateAsync(ServiceTypeRequest request)
		{
			ServiceTypeValidator validator = new();
			var serviceType = new ServiceType(request.Title, request.Icon, request.Color, request.TicketPrefix, request.TargetQueueId);

			if (!validator.Validate(serviceType).IsValid)
				return Unsuccessful<Guid>();

			_repository.Add(serviceType);

			if (!await _repository.UnitOfWork.CommitAsync())
				return Unsuccessful<Guid>(Errors.ErrorSavingDatabase);

			return Successful(serviceType.Id);
		}

		public async Task<Result> DisableAsync(Guid id)
		{
			var serviceType = await _repository.GetByIdAsync(id);
			if(serviceType is null)
				return Unsuccessful(Errors.ServiceTypeNotFound);

			serviceType.ToggleStatus();

			if (!await _repository.UnitOfWork.CommitAsync())
				return Unsuccessful(Errors.ErrorSavingDatabase);

			return Successful();
		}

		public async Task<(Result result, List<ServiceTypeSummary> data)> GetActiveServicesAsync()
		{
			var serviceTypes = await _queries.GetActiveServicesAsync();
			return Successful(serviceTypes);
		}

		public async Task<(Result result, List<ServiceTypeSummary> data)> GetAllAsync()
		{
			var serviceTypes = await _queries.GetListAsync();
			return Successful(serviceTypes);
		}

		public async Task<(Result result, ServiceTypeView data)> GetByIdAsync(Guid id)
		{
			var serviceType = await _queries.GetByIdAsync(id);
			if (serviceType is null)
				return Unsuccessful<ServiceTypeView>(Errors.ServiceTypeNotFound);

			return Successful(serviceType);
		}

		public async Task<Result> ToggleStatusAsync(Guid id)
		{
			var serviceType = await _repository.GetByIdAsync(id);
			if (serviceType is null)
				return Unsuccessful(Errors.ServiceTypeNotFound);

			serviceType.ToggleStatus();

			return Successful();
		}

		public async Task<Result> UpdateAsync(Guid id, ServiceTypeRequest request)
		{
			var serviceType = await _repository.GetByIdAsync(id);
			if(serviceType is null)
				return Unsuccessful(Errors.ServiceTypeNotFound);

			serviceType.Update(request.Title, request.Icon, request.Color, request.TargetQueueId);

			ServiceTypeValidator validator = new();
			var validationResult = validator.Validate(serviceType);

			if (!validationResult.IsValid)
			{
				Notify(validationResult);
				return Unsuccessful();
			}
			
			if (!await _repository.UnitOfWork.CommitAsync())
				return Unsuccessful(Errors.ErrorSavingDatabase);

			return Successful();
		}
	}
}

File: src/Web/src/services/ServiceTypeService.ts
import type { ServiceResult } from "../models/baseServiceModels";
import type {
  ServiceTypeRequest,
  ServiceTypeSummary,
  ServiceTypeView
} from "../models/ServiceTypeModels";
import { BaseService } from "./BaseService";
import type { IServiceTypeService } from "./interfaces/IServiceTypeService";

class ServiceTypeService extends BaseService implements IServiceTypeService {
  private readonly _endpoint = "/totem/ServiceType";

  public async getByIdAsync(id: string): Promise<ServiceResult<ServiceTypeView>> {
    return this.GetAsync<ServiceTypeView>(`${this._endpoint}/${id}`);
  }

  public async getActiveServicesAsync(): Promise<ServiceResult<ServiceTypeSummary[]>> {
    return this.GetAsync<ServiceTypeSummary[]>(`${this._endpoint}/active`);
  }

  public async getAllAsync(): Promise<ServiceResult<ServiceTypeSummary[]>> {
    return this.GetAsync<ServiceTypeSummary[]>(this._endpoint);
  }

  public async createAsync(request: ServiceTypeRequest): Promise<ServiceResult<string>> {
    return this.PostAsync<string, ServiceTypeRequest>(this._endpoint, request, true);
  }

  public async updateAsync(id: string, request: ServiceTypeRequest): Promise<ServiceResult<void>> {
    return this.PutAsync<void, ServiceTypeRequest>(`${this._endpoint}/${id}`, request);
  }

  public async toggleStatusAsync(id: string): Promise<ServiceResult<void>> {
    return this.PatchAsync<void, undefined>(`${this._endpoint}/${id}/toggle-status`, undefined);
  }

  public async disableAsync(id: string): Promise<ServiceResult<void>> {
    return this.PatchAsync<void, undefined>(`${this._endpoint}/${id}/disable`, undefined);
  }

  public async deleteAsync(id: string): Promise<ServiceResult<void>> {
    return this.DeleteAsync<void>(`${this._endpoint}/${id}`);
  }
}

export const serviceTypeService = new ServiceTypeService();
